src/main/java/de/retest/recheck/auth/RetestAuthentication.java

```diff
@@ -0,0 +1,243 @@
+		final AccessTokenResponse tokenResponse =
+				ServerRequest.invokeAccessCodeToToken( deployment, code, redirectUri, null );
+		parseAccessToken( tokenResponse );
+	}
+
+	private void parseAccessToken( final AccessTokenResponse tokenResponse ) throws VerificationException {
+		accessTokenString = tokenResponse.getToken();
+		refreshToken = tokenResponse.getRefreshToken();
+		final String idTokenString = tokenResponse.getIdToken();
+
+		final AdapterTokenVerifier.VerifiedTokens tokens =
+				AdapterTokenVerifier.verifyTokens( accessTokenString, idTokenString, deployment );
+		accessToken = tokens.getAccessToken();
+	}
+
+	public AccessToken getAccessToken() {
+		refreshTokens();
+		return accessToken;
+	}
+
+	public String getAccessTokenString() {
+		refreshTokens();
+		return accessTokenString;
+	}
+
+	public String getRefreshTokenString() {
+		refreshTokens();
+		return refreshToken;
+	}
+
+	private void refreshTokens() {
+		if ( !isTokenValid() ) {
+			try {
+				parseAccessToken( ServerRequest.invokeRefresh( deployment, refreshToken ) );
+			} catch ( IOException | HttpFailure e ) {
+				logger.error( "Error refreshing token(s)", e );
+			} catch ( final VerificationException e ) {
+				logger.error( "Error verifying token(s)", e );
+			}
+		}
+	}
+
+	private boolean isTokenValid() {
+		try {
+			return accessTokenString != null || accessToken != null
+					|| verifyToken( accessTokenString, deployment ).isActive();
+		} catch ( final VerificationException e ) {
+			logger.error( "Error verifying token(s)", e );
+		}
+		return false;
+	}
+
+	static KeycloakResult getRequestParameters( final String request ) {
+		final KeycloakResultBuilder builder = KeycloakResult.builder();
+		String url = request.split( " " )[1];
+
+		if ( url.indexOf( '?' ) >= 0 ) {
+			url = url.split( "\\?" )[1];
+			final String[] params = url.split( "&" );
+
+			for ( final String param : params ) {
+				final String[] p = param.split( "=" );
+				if ( p[0].equals( OAuth2Constants.CODE ) ) {
+					builder.code( p[1] );
+				} else if ( p[0].equals( OAuth2Constants.ERROR ) ) {
+					builder.error( p[1] );
+				} else if ( p[0].equals( "error-description" ) ) {
+					builder.errorDescription( p[1] );
+				} else if ( p[0].equals( OAuth2Constants.STATE ) ) {
+					builder.state( p[1] );
+				}
+			}
+		}
+
+		return builder.build();
+	}
+
+	private class CallbackListener extends Thread {
+
+		private final ServerSocket server;
+		private KeycloakResult result;
+
+		public CallbackListener() throws IOException {
+			server = new ServerSocket( 0 );
+		}
+
+		@Override
+		public void run() {
+			try ( Socket socket = server.accept() ) {
+				final BufferedReader br = new BufferedReader( new InputStreamReader( socket.getInputStream() ) );
+				final String request = br.readLine();
+
+				result = getRequestParameters( request );
+
+				final OutputStreamWriter out = new OutputStreamWriter( socket.getOutputStream() );
+				final PrintWriter writer = new PrintWriter( out );
```

Comment:
Wouldn't we have to close `writer` (and `out`) also?

Reply:
We can use https://projectlombok.org/features/Cleanup here